nav2_costmap_2d/src/layered_costmap.cpp

```diff
@@ -145,40 +145,46 @@ void LayeredCostmap::updateMap(double robot_x, double robot_y, double robot_yaw)
   minx_ = miny_ = std::numeric_limits<double>::max();
   maxx_ = maxy_ = std::numeric_limits<double>::lowest();
 
-  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin();
-    plugin != plugins_.end(); ++plugin)
+  #pragma omp parallel reduction(min:minx_) reduction(min: miny_) \
+  reduction(max: maxx_) reduction(max: maxy_) num_threads(2)
   {
-    double prev_minx = minx_;
-    double prev_miny = miny_;
-    double prev_maxx = maxx_;
-    double prev_maxy = maxy_;
-    (*plugin)->updateBounds(robot_x, robot_y, robot_yaw, &minx_, &miny_, &maxx_, &maxy_);
-    if (minx_ > prev_minx || miny_ > prev_miny || maxx_ < prev_maxx || maxy_ < prev_maxy) {
-      RCLCPP_WARN(
-        rclcpp::get_logger(
-          "nav2_costmap_2d"), "Illegal bounds change, was [tl: (%f, %f), br: (%f, %f)], but "
-        "is now [tl: (%f, %f), br: (%f, %f)]. The offending layer is %s",
-        prev_minx, prev_miny, prev_maxx, prev_maxy,
-        minx_, miny_, maxx_, maxy_,
-        (*plugin)->getName().c_str());
+  #pragma omp for schedule(static, 1)
```

Comment:
From my understanding, this scheduling makes it so that each thread takes 1 iteration at a time statically. Why specify this and not have it use `auto` or the other options? Genuinely not sure, this might be the right thing to do but asking

Reply:
Similar to the previous point, if you have a small number of iteration and `auto` sets a schedule for one thread to take all iterations at once this wouldn't make any sense. I'm just cautious about `auto` because I don't know what the compiler will do here in that case. Therefore one plugin or filter update allocated to one thread seems most appropriate because the number is usually less than 5

Reply:
did you see a demonstrable change for your testing with and without this?

Reply:
i'll get back with some quantitative results on this one, haven't tested with `auto` actually

Reply:
(or even just lacking of schedule specification, which I assume is auto.  I think it defaults to the number of cores)

Reply:
Sooo, first things first i've tried switching the schedule to `auto` and it is performing somewhat the same as with static chunk of size one. avg - average in seconds, first-line - key testing conditions:
```
# schedule(static, 1) num_threads(2)
# avg: 0.002816783901452364 from 12187 samples

# schedule(auto) num_threads(2)
# avg: 0.0027575583444915284 from 6425 samples
```
The difference is barely visible, but I would trust the compiler more than myself in this case therefore all following experiments were done with auto.

In order to see the performance with at least one filter, i've added keep out zones in the setup so map update is slightly longer compared to previous ones. This time experimenting with thread number:
```
# schedule(auto) num_threads(2)
# avg: 0.0031912964961487793 from 5323 samples

# int n_cores = omp_get_num_procs();
# n_threads = std::min(n_cores, (int)std::max(plugins_.size(), filters_.size()));
# schedule(auto) num_threads(n_threads)
# avg: 0.0035948433114727134 from 6084 samples

# int n_cores = omp_get_num_procs();
# int n_threads = std::min(n_cores, (int)std::max(plugins_.size(), filters_.size()));
# int n_plugin_threads = std::min((int)plugins_.size(), n_cores);
# int n_filter_threads = std::min((int)filters_.size(), n_cores);
# avg: 0.0032808737608167084 from 6171 samples
```
First one - same setup as previously + keep-out zones, the second one - single `parallel` block to process (plugins+filter) in the same block with `n_threads` as specified, the third one - two blocks for each `for` loop selecting `n_threads` as shown (two parallel blocks because `omp for` does not allow for `num_threads` option). From this, we see that the second one is lagging behind and the remaining two have quite similar execution duration.

Reply:
Got it - so it sounds like we should leave the scheduling to OpenMP, is it worth the core computations in your opinion? ~90 microseconds difference between them is not really worth optimizing for. 

---

From these results, with the final decision on both of those, what's the compute time speed up as seen by the end user (and what's your compute platform, what layers do you have enabled, etc)

Reply:
Yeah, I think we can leave scheduling for OpenMP. 

I would go with the third one since the difference is not noticeable and it seems more flexible. I've pushed code with that version. Probably those variables in front of the loop could be made static member variables.

I'll update you on the speed-up in the near future.

Reply:
I've got back to the speed-up evaluation. Funny enough, without `DEBUG` log level parallel version is actually underperforming compared to the plain loop... Log level was influencing my measurements previously somehow.

```
plain average update time: 0.001102279375405933
parallel average update time: 0.0010157955731191883
```
These are averaged updated from both local and global cost maps. With this config: https://github.com/simutisernestas/navigation2/blob/parallel_config/nav2_bringup/bringup/params/nav2_params.yaml. I wonder if that would be the case with heavier cost map layers or bigger map.

Reply:
1.1 ms, why so fast? Oh, is this in the TB3 sandbox world? Yeah, it might be worth trying something larger. This shows an overview of getting nav2 running in the AWS warehouse world https://github.com/ros-planning/navigation2/tree/main/nav2_simple_commander.

Reply:
Strangely enough, map updates are even faster on AWS warehouse world demo. I've pulled the latest changes from the repo the code must have been upgraded since the last time I've benchmarked this. Both parallel and serial versions do the job in about the same amount of time. With parallel slightly better, but it's insignificant I would say. I'm testing on `Intel® Core™ i7-3770 CPU @ 3.40GHz × 8` with the same config as mentioned previously.
```
serial avg update time: 0.7053497955298016 ms from 2416 samples
parallel avg update time: 0.6474942507522566 ms from 2991 samples

Reply:
That doesn't surprise me, since you if each thread is doing "more" work, then the overhead of spinning up a new thread is more worth it. It's about 8% it seems, do you know why its lower than what you were seeing before when you reported a ~30% increase in speed? I'm also a little dubious since the updates are so fast. Your other experiment reported 12-20ms in update times (I assume more data) while this is less than 1ms. 1ms seems way too fast to be what you'd see in a practical deployed application so I wonder if in a more realistic setup where things are taking longer if we do see that benefit again.